xarray/core/variable.py

```diff
@@ -376,29 +384,121 @@ def _item_key_to_tuple(self, key):
+            if isinstance(k, Variable):
+                if len(k.dims) > 1:
+                    return self._broadcast_indexes_advanced(key)
+                dims.append(k.dims[0])
+            if not isinstance(k, integer_types):
+                dims.append(d)
+
+        if len(set(dims)) == len(dims):
+            return self._broadcast_indexes_outer(key)
+
+        return self._broadcast_indexes_advanced(key)
+
+    def _broadcast_indexes_basic(self, key):
+        dims = tuple(dim for k, dim in zip(key, self.dims)
+                     if not isinstance(k, integer_types))
+        return dims, BasicIndexer(key)
+
+    def _broadcast_indexes_outer(self, key):
+        dims = tuple(k.dims[0] if isinstance(k, Variable) else dim
+                     for k, dim in zip(key, self.dims)
+                     if not isinstance(k, integer_types))
+        indexer = []
+        for k in key:
+            if isinstance(k, Variable):
+                k = k.data
+
+            if isinstance(k, integer_types + (slice,)):
+                indexer.append(k)
+            else:
+                k = np.asarray(k)
+                if k.ndim > 1:
+                    raise IndexError("Unlabelled multi-dimensional array "
+                                     "cannot be used for indexing.")
+                indexer.append(k if k.dtype.kind != 'b' else k.nonzero()[0])
+        return dims, OuterIndexer(indexer)
+
+    def nonzero(self):
+        """ Equivalent numpy's nonzero but returns a tuple of Varibles. """
+        # TODO we should replace dask's native nonzero
+        # after https://github.com/dask/dask/issues/1076 is implemented.
+        nonzeros = np.nonzero(self.data)
+        return tuple([as_variable(nz, name=dim) for nz, dim
+                      in zip(nonzeros, self.dims)])
+
+    def _broadcast_indexes_advanced(self, key):
+        variables = []
+
+        for dim, value in zip(self.dims, key):
+            if isinstance(value, slice):
+                value = np.arange(*value.indices(self.sizes[dim]))
+
+            try:
+                variable = as_variable(value, name=dim)
+            except MissingDimensionsError:  # change to better exception
+                raise IndexError("Unlabelled multi-dimensional array "
+                                 "cannot be used for indexing.")
+
+            if variable.dtype.kind == 'b':  # boolean indexing case
+                if variable.ndim > 1:
+                    raise IndexError("{}-dimensional boolean indexing is "
+                                     "not supported. ".format(variable.ndim))
+                variables.extend(list(variable.nonzero()))
+            else:
+                variables.append(variable)
+        variables = _broadcast_compat_variables(*variables)
+        dims = variables[0].dims  # all variables have the same dims
+        # overwrite if there is integers
+        key = VectorizedIndexer(k if isinstance(k, integer_types)
```

Comment:
Why is this needed? Is it still helpful now that we have the `OuterIndexer` path?